File: dotnet/src/Microsoft.AutoGen/Core/MessageDelivery.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// MessageDelivery.cs

using System.Reflection;
using Microsoft.AutoGen.Contracts;

namespace Microsoft.AutoGen.Core;

internal sealed class MessageDelivery(MessageEnvelope message, Func<MessageEnvelope, CancellationToken, ValueTask> servicer, IResultSink<object?> resultSink)
{
    public MessageEnvelope Message { get; } = message;
    public Func<MessageEnvelope, CancellationToken, ValueTask> Servicer { get; } = servicer;
    public IResultSink<object?> ResultSink { get; } = resultSink;

    public ValueTask<object?> Future => this.ResultSink.Future;
    public ValueTask FutureNoResult => this.ResultSink.FutureNoResult;

    public ValueTask InvokeAsync(CancellationToken cancellation)
    {
        return this.Servicer(this.Message, cancellation);
    }
}

internal sealed class MessageEnvelope
{
    public object Message { get; }
    public string MessageId { get; }
    public TopicId? Topic { get; private set; }
    public AgentId? Sender { get; private set; }
    public AgentId? Receiver { get; private set; }
    public CancellationToken Cancellation { get; }

    public MessageEnvelope(object message, string? messageId = null, CancellationToken cancellation = default)
    {
        this.Message = message;
        this.MessageId = messageId ?? Guid.NewGuid().ToString();
        this.Cancellation = cancellation;
    }

    public MessageEnvelope WithSender(AgentId? sender)
    {
        this.Sender = sender;
        return this;
    }

    public MessageDelivery ForSend(AgentId receiver, Func<MessageEnvelope, CancellationToken, ValueTask<object?>> servicer)
    {
        this.Receiver = receiver;

        ResultSink<object?> resultSink = new ResultSink<object?>();
        Func<MessageEnvelope, CancellationToken, ValueTask> boundServicer = async (envelope, cancellation) =>
        {
            try
            {
                object? result = await servicer(envelope, cancellation);
                resultSink.SetResult(result);
            }
            catch (TargetInvocationException ex) when (ex.InnerException is OperationCanceledException innerOCEx)
            {
                resultSink.SetCancelled(innerOCEx);
            }
            catch (Exception ex)
            {
                resultSink.SetException(ex);
            }
        };

        return new MessageDelivery(this, boundServicer, resultSink);
    }

    public MessageDelivery ForPublish(TopicId topic, Func<MessageEnvelope, CancellationToken, ValueTask> servicer)
    {
        this.Topic = topic;

        ResultSink<object?> waitForPublish = new ResultSink<object?>();
        Func<MessageEnvelope, CancellationToken, ValueTask> boundServicer = async (envelope, cancellation) =>
        {
            try
            {
                await servicer(envelope, cancellation);
                waitForPublish.SetResult(null);
            }
            catch (Exception ex) // Do we want to special-case cancellation, and hoist the exception type like above?
            {
                waitForPublish.SetException(ex);
            }
        };

        return new MessageDelivery(this, boundServicer, waitForPublish);
    }
}

File: typescript/src/core/MessageDelivery.ts
import { AgentId, TopicId } from "../contracts/IAgentRuntime";
import { ResultSink } from "./ResultSink";
import { MessageContext } from "../contracts/MessageContext";

/**
 * Represents a message being sent through the runtime system.
 */
export class MessageEnvelope {
  private _topic?: TopicId;
  private _sender?: AgentId;
  private _receiver?: AgentId;

  /**
   * Creates a new message envelope.
   * @param message The message content
   * @param messageId Optional unique identifier for the message. If not provided, a new UUID will be generated
   * @param cancellation Optional cancellation signal
   */
  constructor(
    public readonly message: unknown,
    public readonly messageId: string = crypto.randomUUID(),
    public readonly cancellation?: AbortSignal
  ) {}

  /**
   * Sets the sender of the message.
   * @param sender The agent sending the message
   * @returns This message envelope instance for method chaining
   */
  withSender(sender?: AgentId): MessageEnvelope {
    this._sender = sender;
    return this;
  }

  /**
   * Prepares the message for sending to a specific agent.
   * @param receiver The agent that should receive the message
   * @param servicer Function that handles the actual message delivery
   * @returns A new MessageDelivery instance configured for sending
   */
  forSend(
    receiver: AgentId,
    servicer: (envelope: MessageEnvelope, cancellation?: AbortSignal) => Promise<unknown>
  ): MessageDelivery {
    this._receiver = receiver;

    const resultSink = new ResultSink<unknown>();
    const boundServicer = async (envelope: MessageEnvelope, cancellation?: AbortSignal) => {
      try {
        const result = await servicer(envelope, cancellation);
        resultSink.setResult(result);
      } catch (error) {
        resultSink.setError(error instanceof Error ? error : new Error(String(error)));
      }
    };

    return new MessageDelivery(this, boundServicer, resultSink);
  }

  /**
   * Prepares the message for publishing to a topic.
   * @param topic The topic to publish to
   * @param servicer Function that handles the actual message publishing
   * @returns A new MessageDelivery instance configured for publishing
   */
  forPublish(
    topic: TopicId,
    servicer: (envelope: MessageEnvelope, cancellation?: AbortSignal) => Promise<void>
  ): MessageDelivery {
    this._topic = topic;
    return new MessageDelivery(this, servicer);
  }

  /** Gets the topic this message is being published to, if any */
  get topic(): TopicId | undefined {
    return this._topic;
  }

  /** Gets the sender of this message, if any */
  get sender(): AgentId | undefined {
    return this._sender;
  }

  /** Gets the receiver of this message, if any */
  get receiver(): AgentId | undefined {
    return this._receiver;
  }
}

/**
 * Represents a message delivery operation, including the message and its handling mechanism.
 */
export class MessageDelivery {
  /**
   * Creates a new message delivery.
   * @param message The message envelope to deliver
   * @param servicer The function that performs the actual delivery
   * @param resultSink Optional sink for capturing the delivery result
   */
  constructor(
    public readonly message: MessageEnvelope,
    private readonly servicer: (envelope: MessageEnvelope, cancellation?: AbortSignal) => Promise<void>,
    private readonly resultSink?: ResultSink<unknown>
  ) {}

  /** Gets the future that will resolve with the delivery result */
  get future(): Promise<unknown> {
    return this.resultSink?.future ?? Promise.resolve(null);
  }

  /**
   * Executes the message delivery.
   * @param cancellation Optional cancellation signal
   */
  async invokeAsync(cancellation?: AbortSignal): Promise<void> {
    await this.servicer(this.message, cancellation);
  }
}
